src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,262 @@
+        {
+            foreach (var info in localFuncInfos)
+            {
+                info.IsDirty = false;
+            }
+        }
+
+        /// <summary>
+        /// The results from running the assignment pass.
+        /// </summary>
+        protected virtual LocalFuncAssignmentResults GetResults() => null;
+
+        /// <summary>
+        /// Records all possibly-unassigned reads in all local
+        /// functions in the given block.
+        /// </summary>
+        private void RecordReadsInLocalFunctions(BoundBlock block)
+        {
+            if (block.LocalFunctions.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+
+            var recordingState = _recordingReads;
+            _recordingReads = true;
+
+            do
+            {
+                ClearDirtyBits(_readVars.Values);
+
+                foreach (var stmt in block.Statements)
+                {
+                    if (stmt.Kind == BoundKind.LocalFunctionStatement)
+                    {
+                        var savedState = this.State;
+                        var localFunc = (BoundLocalFunctionStatement)stmt;
+
+                        // Clear the state before visitation since all captured
+                        // reads could be unassigned at the local function callsite
+                        this.State = new LocalState(
+                            BitVector.Create(this.State.Assigned.Capacity));
+
+                        // Assign all parameters of all parent functions
+                        EnterAllParameters();
+
+                        VisitLambdaOrLocalFunction(localFunc,
+                            recordAssigns: false);
+
+                        this.State = savedState;
+                    }
+                }
+            } while (HasDirtyLocalFunctions(_readVars.Values));
+
+            _recordingReads = recordingState;
+        }
+
+        private void EnterAllParameters()
+        {
+            foreach (var info in variableBySlot)
+            {
+                if (info.Symbol?.Kind == SymbolKind.Parameter)
+                {
+                    EnterParameter((ParameterSymbol)info.Symbol);
+                }
+            }
+        }
+
+        private void RecordReadInLocalFunction(int slot)
+        {
+            var localFunc = GetNearestLocalFunctionOpt(currentMethodOrLambda);
+
+            Debug.Assert(localFunc != null);
+
+            LocalFuncInfo readInfo;
+            if (_readVars.TryGetValue(localFunc, out readInfo))
+            {
+                if (!readInfo.UsedVars.Contains(slot))
+                {
+                    readInfo.IsDirty = true;
+                    readInfo.UsedVars.Add(slot);
+                }
+            }
+            else
+            {
+                readInfo = new LocalFuncInfo();
+                readInfo.IsDirty = true;
+                readInfo.UsedVars.Add(slot);
+                _readVars.Add(localFunc, readInfo);
+            }
```

Comment:
Consider extracting a `RecordLocalFunction(_readVars, localFunc, slot)` helper method for use here and in `RecordIfCaptured`.